File: src/main/java/com/muxiu1997/sharewhereiam/loader/NetworkLoader.java
package com.muxiu1997.sharewhereiam.loader;

import com.muxiu1997.sharewhereiam.network.MessageShareWaypoint;
import com.muxiu1997.sharewhereiam.network.MessageStartShareWaypoint;
import com.muxiu1997.sharewhereiam.network.NetworkHandler;
import cpw.mods.fml.common.network.simpleimpl.IMessage;
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler;
import cpw.mods.fml.relauncher.Side;

public class NetworkLoader {

    public static final NetworkLoader INSTANCE = new NetworkLoader();

    public void load() {
        registerMessage(MessageStartShareWaypoint.Handler.class, MessageStartShareWaypoint.class);
        registerMessage(MessageShareWaypoint.Handler.class, MessageShareWaypoint.class);
    }

    private static int nextID = 0;

    private static <REQ extends IMessage, REPLY extends IMessage> void registerMessage(Class<? extends IMessageHandler<REQ, REPLY>> messageHandler, Class<REQ> requestMessageType) {
        NetworkHandler.INSTANCE.registerMessage(messageHandler, requestMessageType, nextID++, Side.SERVER);
        NetworkHandler.INSTANCE.registerMessage(messageHandler, requestMessageType, nextID++, Side.CLIENT);
    }
}

File: src/main/kotlin/com/muxiu1997/sharewhereiam/loader/NetworkLoader.kt
package com.muxiu1997.sharewhereiam.loader

import com.muxiu1997.sharewhereiam.network.MessageShareWaypoint
import com.muxiu1997.sharewhereiam.network.MessageStartShareWaypoint
import com.muxiu1997.sharewhereiam.network.network
import cpw.mods.fml.common.network.simpleimpl.IMessage
import cpw.mods.fml.common.network.simpleimpl.IMessageHandler
import cpw.mods.fml.relauncher.Side

object NetworkLoader {
    fun load() {
        registerServerSide(MessageStartShareWaypoint.Companion.Handler)
        registerClientSide(MessageStartShareWaypoint.Companion.Handler)
        registerClientSide(MessageShareWaypoint.Companion.Handler)
    }

    private inline fun <reified REQ : IMessage?, REPLY : IMessage?> registerServerSide(
        messageHandler: IMessageHandler<REQ, REPLY>,
    ) {
        network.registerMessage(messageHandler, REQ::class.java, nextID++, Side.SERVER)
    }

    private inline fun <reified REQ : IMessage?, REPLY : IMessage?> registerClientSide(
        messageHandler: IMessageHandler<REQ, REPLY>,
    ) {
        network.registerMessage(messageHandler, REQ::class.java, nextID++, Side.CLIENT)
    }

    private var nextID = 0
}
